### PR TITLE
Feature/progress bar cb

### PR DESCRIPTION
**Pull request by zwiersj003**

### CODE REVIEW

File: ISX/ISX-CabinCrew/CabinCrewViewController.swift [hunk 1]
```diff
@@ -110,6 +116,35 @@ class CabinCrewViewController: UITableViewController {
         }
     }
     
+    /*
+     Returns array with flights
+     */
+    func flights() {
```
Comment (robbym1991) [inline ISX/ISX-CabinCrew/CabinCrewViewController.swift]: duidelijkere methodenaame

File: ISX/ISX-CabinCrew/CabinCrewViewController.swift [hunk 2]
```diff
@@ -110,6 +116,35 @@ class CabinCrewViewController: UITableViewController {
...
+        flightsRef?.observe(.value, with: { snapshot in
+            for item in snapshot.children {
+                if let flight = item as? DataSnapshot {
+                    let flightnumber = Flight(snapshot: flight)
+                        self.flightsArray.append(flightnumber)
```
Comment (robbym1991) [inline ISX/ISX-CabinCrew/CabinCrewViewController.swift]: tab te ver

File: ISX/ISX-CabinCrew/CabinCrewViewController.swift [hunk 3]
```diff
@@ -110,6 +116,35 @@ class CabinCrewViewController: UITableViewController {
...
+                }
+            }
+        })
+    }
+    
+    /*
+     Returns array with products
+    */
+    func products() {
```
Comment (robbym1991) [inline ISX/ISX-CabinCrew/CabinCrewViewController.swift]: methodnaam kan beter

File: ISX/ISX-CabinCrew/CabinCrewViewController.swift [hunk 4]
```diff
@@ -120,23 +155,45 @@ class CabinCrewViewController: UITableViewController {
         return 130
```
Comment (robbym1991) [inline ISX/ISX-CabinCrew/CabinCrewViewController.swift]: magic number

Comment (robbym1991) [reply ISX/ISX-CabinCrew/CabinCrewViewController.swift]: bijvoorbeeld beter om in Constants te verwerken

File: ISX/ISX-CabinCrew/CabinCrewViewController.swift [hunk 5]
```diff
@@ -120,23 +155,45 @@ class CabinCrewViewController: UITableViewController {
...
     }
     
-     override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+    override func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
+        guard let progressBarCell = tableView.dequeueReusableCell(withIdentifier: "progressCell") as? ProgressBarCell
+            else { return UITableViewCell() }
+        if (flightsArray.count > 0 ){
```
Comment (robbym1991) [inline ISX/ISX-CabinCrew/CabinCrewViewController.swift]: spatie ){

File: ISX/ISX-CabinCrew/CabinCrewViewController.swift [hunk 6]
```diff
@@ -15,11 +15,14 @@ import NVActivityIndicatorView
...
     var requestsArray = [Request]()
+    var flightsArray = [Flight]()
+    var productsArray = [Product]()
     private var datarootRef: DatabaseReference?
     private var requestsRef: DatabaseReference?
     private var productsRef: DatabaseReference?
+    private var flightsRef: DatabaseReference?
     var activityIndicatorView: NVActivityIndicatorView?
-
+
```
Comment (Maren-Osnabrug) [inline ISX/ISX-CabinCrew/CabinCrewViewController.swift]: whitespace

File: ISX/ISX-CabinCrew/CabinCrewViewController.swift [hunk 7]
```diff
@@ -40,10 +43,13 @@ class CabinCrewViewController: UITableViewController {
         datarootRef = Database.database().reference(withPath: "dataroot")
         requestsRef = datarootRef?.child("requests")
         productsRef = datarootRef?.child("products")
+        flightsRef = datarootRef?.child("flights")
         
         requestsRef?.keepSynced(true)
         observeRequests()
         observeNewRequest()
+        flights()
```
Comment (Maren-Osnabrug) [inline ISX/ISX-CabinCrew/CabinCrewViewController.swift]: these aren't descriptive method names

File: ISX/ISX-CabinCrew/CabinCrewViewController.swift [hunk 8]
```diff
@@ -40,10 +43,13 @@ class CabinCrewViewController: UITableViewController {
...
+        products()
```
Comment (Maren-Osnabrug) [inline ISX/ISX-CabinCrew/CabinCrewViewController.swift]: same as above

File: ISX/ISX-CabinCrew/CabinCrewViewController.swift [hunk 9]
```diff
@@ -110,6 +116,35 @@ class CabinCrewViewController: UITableViewController {
...
     }
     
+    /*
+     Returns array with flights
+     */
+    func flights() {
+        flightsRef?.observe(.value, with: { snapshot in
+            for item in snapshot.children {
+                if let flight = item as? DataSnapshot {
+                    let flightnumber = Flight(snapshot: flight)
```
Comment (Maren-Osnabrug) [inline ISX/ISX-CabinCrew/CabinCrewViewController.swift]: you are appending a flight, not a number

File: ISX/ISX-CabinCrew/CabinCrewViewController.swift [hunk 10]
```diff
@@ -110,6 +116,35 @@ class CabinCrewViewController: UITableViewController {
...
+    }
+    
+    /*
+     Returns array with products
+    */
+    func products() {
+        productsRef?.observe(.value, with: { snapshot in
+            for item in snapshot.children {
+                if let product = item as? DataSnapshot {
+                    let products = Product(snapshot: product)
```
Comment (Maren-Osnabrug) [inline ISX/ISX-CabinCrew/CabinCrewViewController.swift]: you are appending one product

File: ISX/ISX-CabinCrew/CabinCrewViewController.swift (hunk 4 above)
Comment (Maren-Osnabrug) [inline ISX/ISX-CabinCrew/CabinCrewViewController.swift]: use constant

File: ISX/ISX-CabinCrew/CabinCrewViewController.swift [hunk 11]
```diff
@@ -120,23 +155,45 @@ class CabinCrewViewController: UITableViewController {
         return 130
     }
     
-     override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+    override func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
+        guard let progressBarCell = tableView.dequeueReusableCell(withIdentifier: "progressCell") as? ProgressBarCell
+            else { return UITableViewCell() }
+        if (flightsArray.count > 0 ){
+            let flight = flightsArray[1]
```
Comment (Maren-Osnabrug) [inline ISX/ISX-CabinCrew/CabinCrewViewController.swift]: why the second flight

File: ISX/ISX-CabinCrew/CabinCrewViewController.swift [hunk 12]
```diff
@@ -120,23 +155,45 @@ class CabinCrewViewController: UITableViewController {
...
+                }
+            }
+            progressBarCell.setProgressBar(flight: flight, currentRevenue: currentRevenue)
+            progressBarCell.isUserInteractionEnabled = false
+        }
+        return progressBarCell
+    }
+    
+    override func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
+        return 100
```
Comment (Maren-Osnabrug) [inline ISX/ISX-CabinCrew/CabinCrewViewController.swift]: constant

File: ISX/ISX-CabinCrew/CabinCrewViewController.swift [hunk 13]
```diff
@@ -120,23 +155,45 @@ class CabinCrewViewController: UITableViewController {
...
+    }
+    
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
         guard let cell = tableView.dequeueReusableCell(withIdentifier: "CustomRequestCell", for: indexPath) as? RequestCell else { return UITableViewCell() }
         let request = requestsArray[indexPath.row]
         cell.setCellData(request: request)
-        
+    
         productsRef?.observe(.value, with: { snapshot in
             let arr = snapshot.children.allObjects as NSArray
```
Comment (Maren-Osnabrug) [inline ISX/ISX-CabinCrew/CabinCrewViewController.swift]: array of products?

File: ISX/ISX-CabinCrew/CabinCrewViewController.swift [hunk 14]
```diff
@@ -120,23 +155,45 @@ class CabinCrewViewController: UITableViewController {
...
-            for item in arr {
-                let item = Product(snapshot: item as! DataSnapshot)
-                if item.id == String(request.productId) {
-                    cell.productName.text = item.title
+                for item in arr {
+                    let item = Product(snapshot: item as! DataSnapshot)
+                    if item.id == String(request.productId) {
```
Comment (Maren-Osnabrug) [inline ISX/ISX-CabinCrew/CabinCrewViewController.swift]: brackets

File: ISX/ISX/Cells/ProgressBarCell.swift [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+    @IBOutlet weak var targetRevenueAmountLabel: UILabel!
+    
+    func setProgressBar (flight: Flight, currentRevenue: Int) {
+        var progress: Float
+        if(currentRevenue <= flight.averageRevenue) {
+            progress = (Float(currentRevenue)/Float(flight.highestRevenue))
+            targetRevenueTextLabel.text = "Target Revenue: €\(currentRevenue)/€\(flight.averageRevenue)"
+            targetRevenueAmountLabel.text = "€\(flight.averageRevenue)"
+        } else {
+            progress = (Float(currentRevenue)/Float(flight.highestRevenue))
```
Comment (Maren-Osnabrug) [inline ISX/ISX/Cells/ProgressBarCell.swift]: this line and the one below are the same in both statements, so they can be place above the if/else

File: ISX/ISX/Models/Flight.swift [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+        self.ref = nil
+    }
+    
+    init(snapshot: DataSnapshot) {
+        let dict = snapshot.value as! [String:AnyObject]
+        self.id = dict["id"] as! Int
+        self.destination = dict["destination"] as! String
+        self.averageRevenue = dict["averageRevenue"] as! Int
+        self.highestRevenue = dict["highestRevenue"] as! Int
+        self.ref = snapshot.ref
```
Comment (Maren-Osnabrug) [inline ISX/ISX/Models/Flight.swift:34]: this looks fine, like all the other models, but we might need to consider refactoring without force casting in the future

**Review by Maren-Osnabrug — CHANGES_REQUESTED** [review]

File: ISX/ISX-CabinCrew/CabinCrewViewController.swift (hunk 12 above)
Comment (robbym1991) [inline ISX/ISX-CabinCrew/CabinCrewViewController.swift]: magic number

**Review by robbym1991 — CHANGES_REQUESTED** [review]

**Review by Maren-Osnabrug — APPROVED** [review]

**Review by robbym1991 — APPROVED** [review]

**Pull request merged**